### PR TITLE
Add a rust linter and formatter (Take 2)

### PR DESCRIPTION
**Pull request by Cadair**
This adds some rust specific pre-commit hooks as recommended by @Will-Shanks in #110 

It seems that these checks don't play nice with pre-commit.ci so I have swapped to using pre-commit.ci lite and GH actions to run them, which means enabling autofix.

Oh, also somewhat controversially I have configured clippy to enforce the presence of return statements, as I think that will be better for people likely to contribute here who are more familiar with Python.

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -40,7 +40,15 @@ repos:
...
     hooks:
-        - id: black
+      - id: black
+  # Rust code formatting
+  - repo: https://github.com/FeryET/pre-commit-rust
+    rev: v1.1.0
+    hooks:
+      - id: cargo-check
+      - id: clippy
+        args: ["--allow-staged", "--fix", "--", "-A", "clippy::needless_return"]
```
Comment (nabobalis) [inline .pre-commit-config.yaml]: You didn't want to config this?

Comment (Cadair) [reply .pre-commit-config.yaml]: As far as I can tell there isn't a decent way to config this which works with the slightly old rust I have on my system :p

File: .github/workflows/pre-commit.yml [hunk 2]
```diff
@@ -0,0 +1,20 @@
+name: pre-commit
```
Comment (nabobalis) [inline .github/workflows/pre-commit.yml:1]: The nuclear option.

**Review by nabobalis — APPROVED** [review]

File: .github/workflows/pre-commit.yml (hunk 2 above)
Comment (Cadair) [reply .github/workflows/pre-commit.yml:1]: Indeed.

**Pull request merged**